server/src/main/java/org/elasticsearch/common/blobstore/RetryingInputStream.java

```diff
@@ -0,0 +1,327 @@
+ */
+
+package org.elasticsearch.common.blobstore;
+
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.core.IOUtils;
+import org.elasticsearch.repositories.blobstore.RequestedRangeNotSatisfiedException;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.NoSuchFileException;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.elasticsearch.core.Strings.format;
+
+public abstract class RetryingInputStream extends InputStream {
+
+    private static final Logger logger = LogManager.getLogger(RetryingInputStream.class);
+
+    public static final int MAX_SUPPRESSED_EXCEPTIONS = 10;
+
+    private final BlobStoreServices blobStoreServices;
+    private final OperationPurpose purpose;
+    private final long start;
+    private final long end;
+    private final List<Exception> failures;
+
+    protected SingleAttemptInputStream currentStream;
+    private int attempt = 1;
+    private int failuresAfterMeaningfulProgress = 0;
+    private long currentOffset = 0;
+    private boolean closed = false;
+
+    protected RetryingInputStream(BlobStoreServices blobStoreServices, OperationPurpose purpose) throws IOException {
+        this(blobStoreServices, purpose, 0L, Long.MAX_VALUE - 1L);
+    }
+
+    @SuppressWarnings("this-escape") // TODO: We can do better than this but I don't want to touch the tests for the first implementation
+    protected RetryingInputStream(BlobStoreServices blobStoreServices, OperationPurpose purpose, long start, long end) throws IOException {
+        if (start < 0L) {
+            throw new IllegalArgumentException("start must be non-negative");
+        }
+        if (end < start || end == Long.MAX_VALUE) {
+            throw new IllegalArgumentException("end must be >= start and not Long.MAX_VALUE");
+        }
+        this.blobStoreServices = blobStoreServices;
+        this.purpose = purpose;
+        this.failures = new ArrayList<>(MAX_SUPPRESSED_EXCEPTIONS);
+        this.start = start;
+        this.end = end;
+        final int initialAttempt = attempt;
+        openStreamWithRetry();
+        maybeLogAndRecordMetricsForSuccess(initialAttempt, "open");
+    }
+
+    private void openStreamWithRetry() throws IOException {
+        while (true) {
+            if (currentOffset > 0 || start > 0 || end < Long.MAX_VALUE - 1) {
+                assert start + currentOffset <= end
+                    : "requesting beyond end, start = " + start + " offset=" + currentOffset + " end=" + end;
+            }
+            try {
+                currentStream = blobStoreServices.getInputStream(Math.addExact(start, currentOffset), end);
+                return;
+            } catch (NoSuchFileException | RequestedRangeNotSatisfiedException e) {
+                throw e;
+            } catch (RuntimeException | IOException e) {
```

Comment:
I think the current S3 implementation [does not retry on IOException](https://github.com/elastic/elasticsearch/blob/574efab401fbbcbe59bcd0b66c260067d64545e6/modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java#L106) at opening time?

Reply:
Hmm yes that's a good point. That may be a relic that came about when implementing the other CSPs. I'll make it consistent with the existing impl for now and re-assess the need for that when we add them in.